### ISSUE TITLE
Compile Errors in analytics test suite

### ISSUE THREAD

**Issue by rkapl123** [OpenSourceRisk/Engine#160]
Dear All!

After fetching the latest version (1.8.10.0) using the tag v.1.8.10.0 and the associated Quantlib I got following compile errors on the analytics test suite using Visual Studio 2019 (cmake):

`"C:\dev\Engine\build\ALL_BUILD.vcxproj" (Standardziel) (1) ->
"C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj" (Standardziel) (27) ->
(ClCompile Ziel) ->
  C:\dev\boost\boost/timer.hpp(45,1): error C2869: "boost::timer": Wurde bereits als Namespace definiert (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\dev\E
ngine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(723,1): error C2882: "timer": Unzulässige Verwendung eines Namespace-Bezeichners in einem Ausdruck (Quelldatei wird kompiliert C:\dev\Engine\ORE
Analytics\test\testsuite.cpp) [C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(723,26): error C2146: Syntaxfehler: Fehlendes ";" vor Bezeichner "tu_timer" (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuit
e.cpp) [C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(723,26): error C2065: "tu_timer": nichtdeklarierter Bezeichner (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\de
v\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(735,73): error C2065: "tu_timer": nichtdeklarierter Bezeichner (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\de
v\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(735,1): error C2660: "boost::unit_test::framework::state::child_timeout": Funktion akzeptiert keine 1 Argumente (Quelldatei wird kompiliert C:\d
ev\Engine\OREAnalytics\test\testsuite.cpp) [C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(767,77): error C2065: "tu_timer": nichtdeklarierter Bezeichner (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\de
v\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(767,1): error C2660: "boost::unit_test::framework::state::child_timeout": Funktion akzeptiert keine 1 Argumente (Quelldatei wird kompiliert C:\d
ev\Engine\OREAnalytics\test\testsuite.cpp) [C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(777,55): error C2065: "tu_timer": nichtdeklarierter Bezeichner (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\de
v\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
  C:\dev\boost\boost/test/impl/framework.ipp(790,55): error C2065: "tu_timer": nichtdeklarierter Bezeichner (Quelldatei wird kompiliert C:\dev\Engine\OREAnalytics\test\testsuite.cpp) [C:\de
v\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]`

The rest compiled successful.

-regards,
Roland

**Comment by rkapl123** [comment]
OK, after the compile error still prevails, I've analyzed the situation a bit more.

The problem arised when using the header only version of unit_test.hpp in [testsuite.cpp](https://github.com/OpenSourceRisk/Engine/blob/v1.8.8.0/OREAnalytics/test/testsuite.cpp)
 of version 1.8.8: `#include boost/test/included/unit_test.hpp`
However, in that version there were no time measurements needed in testsuite.cpp, so the effect didn't show until 1.8.10, where 
```
#include <boost/timer/timer.hpp>
using namespace boost;
using boost::timer::cpu_timer;
```
was introduced again, invoking the boost internal incompatibility.
Probably the compile error doesn't show up elsewhere because `ORE_ENABLE_PARALLEL_UNIT_TEST_RUNNER` is defined, thus preventing the offending include:
```
#ifdef ORE_ENABLE_PARALLEL_UNIT_TEST_RUNNER
#include <test-suite/paralleltestrunner.hpp>
#else
#include <boost/test/included/unit_test.hpp>
#endif
```
An easy fix is falling back to the linked version of unit_test.hpp using `#include boost/test/unit_test.hpp`, however I'm not sure whether this is intended anymore, after it's been changed in 1.8.8

-regards,
Roland

**Comment by rkapl123** [comment]
OK, after upgrading to boost 1_78 as suggested by the new userguide (sorry for not doing this before), I've come accross another problem, this time with linking:
```
testsuite.obj : error LNK2038: Konflikt ermittelt für "boost_log_abi": Der Wert "v2s_mt_nt62" stimmt nicht mit dem Wert "v2s_mt_nt6" in amcbermudanswaption.obj überein. [C:\dev\Engine\build
\OREAnalytics\test\orea-test-suite.vcxproj]
     Bibliothek "C:/dev/Engine/build/OREAnalytics/test/Release/orea-test-suite.lib" und Objekt "C:/dev/Engine/build/OREAnalytics/test/Release/orea-test-suite.exp" werden erstellt.
C:\dev\Engine\build\OREAnalytics\test\Release\orea-test-suite.exe : fatal error LNK1319: 1 Konflikte ermittelt [C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj]
Die Erstellung des Projekts "C:\dev\Engine\build\OREAnalytics\test\orea-test-suite.vcxproj" ist abgeschlossen (Standardziele) -- FEHLER.
```
The other parts compile/link without problems.
-regards,
Roland

**Comment by byouness** [comment]
Hello Roland,

Did you manage to fix this issue or work around it?

Best regards,
Youness

**Comment by rkapl123** [comment]
No, but I found a possible reason [here](https://github.com/microsoft/vcpkg/issues/22495) and [here](https://stackoverflow.com/questions/74904263/building-boost-log-library)
It seems that I'm using a boost lib that was built with v2s_mt_nt62 and when compiling amcbermudanswaption, the build uses v2s_mt_nt6. There is a hint to use BOOST_USE_WINAPI_VERSION that I'm currently following...

**Comment by gbfredrik** [comment]
Hi all,
We were recently hit by the same issue and solved it temporarily by adding this line at the top of OREAnalytics/test/testsuite.cpp
`#define _WIN32_WINNT 0x0600`

My colleague was able to set it through the CMake files as well, with something like:
`target_compile_definitions(orea-test-suite PUBLIC _WIN32_WINNT=0x0600)`

The issue seems to be caused by how Boost.log determines the Windows SDK version. I’m hoping someone more knowledgeable on this could help advise on any better solutions. As Roland mentioned, it might be more correct to use the Boost flag `BOOST_USE_WINAPI_VERSION`.

/Fredrik

**Comment by rkapl123** [comment]
So, I think I have a working solution, either fredriks suggestion or the BOOST_USE_WINAPI_VERSION are equivalently doing the same.
What's important: add the target_compile_definitions after the add_executable of the orea-test-suite in $ORE/OREAnalytics/test/CMakeLists.txt:
```
....
add_executable(orea-test-suite ${OREAnalytics-Test_SRC})
target_link_libraries(orea-test-suite ${QL_LIB_NAME})
target_link_libraries(orea-test-suite ${QLE_LIB_NAME})
target_link_libraries(orea-test-suite ${ORED_LIB_NAME})
target_link_libraries(orea-test-suite ${OREA_LIB_NAME})
target_link_libraries(orea-test-suite ${Boost_LIBRARIES} ${RT_LIBRARY})

# target_compile_definitions(orea-test-suite PUBLIC _WIN32_WINNT=0x0600)
target_compile_definitions(orea-test-suite PUBLIC BOOST_USE_WINAPI_VERSION=0x0600)
```

What I'm still struggling is the fact that my platform (win 10) is not sufficient for the "auto-selected" one by boost.log, according to [the site Andrey Semashev referred to](https://learn.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt?view=msvc-170&redirectedfrom=MSDN) this should be 0x0A00:
```
....
#define _WIN32_WINNT_VISTA                  0x0600 // Windows Vista
#define _WIN32_WINNT_WS08                   0x0600 // Windows Server 2008
#define _WIN32_WINNT_WIN7                   0x0601 // Windows 7
#define _WIN32_WINNT_WIN8                   0x0602 // Windows 8
#define _WIN32_WINNT_WINBLUE                0x0603 // Windows 8.1
#define _WIN32_WINNT_WIN10                  0x0A00 // Windows 10
```
But any value above 0x0600 brings back the errors...

**Comment by byouness** [comment]
Hi Roland, Fredrik!
Thanks for your help. That worked for me.
Best,
Youness

**Comment by mgroncki** [comment]
I've noticed that this problem appears using the prebuilt boost binaries 1.78 or later under windows.  But if I build boost from the sources I can compile the test-suite without any error. I've tested it with boost 1.81 with msvc14.3 on x64

**Comment by pcaspers** [comment]
I understand the OREData test suite builds fine? So we should compare the build configurations OREAnalytics / OREData.

We noticed that `log` is not part of the boost required components in `OREAnalytics/CMakeLists.txt`. Could you try adding this and see if it changes anything?

**Comment by rkapl123** [comment]
Well, actually the original compile problem already has been fixed by upgrading the boost lib to 1_78 as suggested by the userguide. We're now struggling with incompatibilities in the prebuilt windows boost libraries, especially starting with 1_78. Mathias suggested it should work with 1_72 in the discussion of PR #193 .
I've not tried to build with 1_72, it also isn't available vor VS 2022 anymore, so I'm rather reluctant to step back.

**Comment by pcaspers** [comment]
Ok, but prebuilt or not, there must be a difference between OREData and OREAnalytics test suites causing the latter to fail while the former works? I guess it would be helpful to identify this difference.

**Comment by rkapl123** [comment]
Well, after checking this, I found that no other test project uses `boost/test/test_tools.hpp` (being only in testsuite.cpp of OREAnalytics test project). 
I guess this (and the usage of boost/timer/timer.hpp as described in https://github.com/OpenSourceRisk/Engine/issues/160#issuecomment-1774041370) cause the boost internal incompatibilities, for which purpose it is included I don't know (from the header doc: This file is used to select the test tools implementation and includes all the necessary headers).

I've uncommented the line `#include <boost/test/test_tools.hpp>` in OREAnalytics\test\testsuite.cpp and it compiled well using boost 1_72.

In general, there are some mismatches in the minimum boost version required, first there is the cmake check that seems to only require 1_58, then there ist the user doc that states "On all platforms the minimum required boost version is 1_78" and finally in the hints for building ORE under windows on page 19, there is "%BOOST% pointing to your directory, e.g, C:\boost_1_72_0".
I'd rather go for a boost version as low as possible and also try to incorporate this in the build tests using the prebuilt windows binaries.

**Comment by mgroncki** [comment]
If  the macro _WIN32_WINNT is not defined during the boost building process, boost will set `BOOST_USE_WINAPI_VERSION` by default to 0x601 (windows 7).Otherwise, it will be set to the value of `_WIN32_WINNT`.

In ORE we don't explicitly define_WIN32_WINNT, so `BOOST_USE_WINAPI_VERSION` should default to Win7. 

But if we include windows headers like `<windows.h>` it will be defined. <windows.h> includes the header `<sdkddkver.h>` where _WIN32_WINNT is set to default value. If one build ORE on Windows 10 it will be defaulted to 0x0A00. 

The include `<boost/test/include/unit_test.hpp>` in testsuite.cpp in the  orea-testsuite includes several files and one of them includes <windows.h> and that defines _WIN32_WINNT. 

If I change the include to <boost/test/unit_test.hpp> (which we use in the other test-suites) there is no namespace conflict in boost log.

**Comment by rkapl123** [comment]
I have also tried the other end of boost and succesfully built ORE with the above mentioned changes (uncommenting `#include <boost/test/test_tools.hpp>` (this solves the first compilation error) and changing `#include boost/test/included/unit_test.hpp` to `#include boost/test/unit_test.hpp` (this solves the ABI problem mentioned later) in OREAnalytics\test\testsuite.cpp) using boost 1_78 and 1_83 (self built and prebuilt).

The lowest version of boost I still could build ORE was 1_69 (also with these changes). The `PUBLIC BOOST_USE_WINAPI_VERSION=0x0600` patch is not necessary anymore here.

So I'd rather also add uncommenting `#include <boost/test/test_tools.hpp>`  to the PR.

**Comment by mgroncki** [comment]
I've just noticed, that the paralleltestrunner.hpp will  also include <windows.h>, so same problem here. So maybe the best way to solve it, is to define the macros _WIN32_WINNT and _WINVER for all libraries. That would avoid any unintended mismatch between the different ORE libraries on Windows.
Any thoughts on that?

**Comment by rkapl123** [comment]
Well, I think that this is actually a Boost internal problem (I had the boost_log_abi error even when compiling boost from source, which still puzzles me quite a bit. Only after changing `#include boost/test/included/unit_test.hpp` to `#include boost/test/unit_test.hpp` this could be resolved).
I know the paralleltestrunner.hpp is actually in the quantlib part (making a similar change a bit problematic), but it seems that it copes with the abi problem, at least I could build OREAnalystics testsuite with ORE_ENABLE_PARALLEL_UNIT_TEST_RUNNER set to ON without problems.
I'd rather avoid setting _WIN32_WINNT and _WINVER to a lower version than the current one, except that Acadia/All other cotributers promise to not use any features found in higher versions. Then we'd have the same issues again.
But I really appreciate @kohnech Initiative to add a windows workflow to the pipeline, this should reduce windows related issues a lot in the future.
I'd also suggest to add a minimum version in cmake (similar to quantlib) and make this consistent with the userdoc, then anybody will immediately fail on building with a too low boost version. Based on my experiences this should either be 1_69 or 1_72, however the userdoc says 1_78, maybe @pcaspers /Acadiasoft can comment on this.

**Comment by mgroncki** [comment]
As I mentioned earlier, if you don't specify BOOST_USE_WINAPI_VERSION or _WIN32_WINNT during the boost build, BOOST_USE_WINAPI_VERSION will be set to 0x0601 (windows 7). 

To change this you need to build boost with `define=BOOST_USE_WINAPI_VERSION=0x0A00` or `define=_WIN32_WINNT=0x0A00` in the b2 command line build tool. 

From 1.78 onwards boost-log will derive the namespace from the value in BOOST_USE_WINAPI_VERSION. 
It will be v2s_mt_nt6 for <=Windows 7  and v2s_mt_nt62 for >=Win8. 
You can check the used namespace with DUMPBIN on the log library:

`dumpbin /all libboost_log-vc143-mt-x64-1_81.lib /out:dump.txt`

If I build boost with boost_use_win_api_version set to win10 and use that in ORE without defining _WIN32_WINNT = 0x0A00, it will result in linker errors in ORE-Data. Which make sense, _WIN32_WINNT is not set and if we include boost-log headers it will default the macro to Win7 but the library was built with the api set to Win10.

I just had a brief look yesterday in the paralleltestrunner and saw uses the header only unit-test, which caused problems. But you are right, we can build the the test-suite with paralleltestrunner  using the prebuilt boost. 

I had look into it and although_WIN32_WINNT will be set to Win10, there is another boost header included before the unit-test which defaults BOOST_USE_WINAPI_VERSION to win7. the macro _WIN32_WINNT will be defined after that. 

Looking at the release notes of boost.178 [ReleaseNotes](https://www.boost.org/users/history/version_1_78_0.html)
> Boost 1.78 General changes: On Windows, when building the library for Windows 8 or later, the library will use nt62 tag in the version namespace to denote the target OS ABI. For example, the version namespace could be named as v2_mt_nt62. This name will be part of all symbols exported by the library. Use the BOOST_USE_WINAPI_VERSION macro consistently when building Boost and your code to request the minimum target Windows version.

I think we should set BOOST_USE_WINAPI_VERSION and _WIN32_WINNT and _WINVER explicit and consistent between all libraries. At the moment its a bit dependent on the order of the includes.

From [MSDN](https://learn.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt?view=msvc-170) 

> :The preprocessor macros WINVER and _WIN32_WINNT specify the minimum operating system version your code supports. Visual Studio and the Microsoft C++ compiler support targeting Windows 7 SP1 and later.

I would like to set it to Win7 to avoid any problems with the current prebuilt boost libraries. That would be the easiest for all users/developers.

And if there is a need to support some newer features of the WINAPI, one can build a custom boost with defining BOOST_USE_WINAPI_VERSION and adjusting the two macros in the ORE project.

**Comment by rkapl123** [comment]
OK, that means we define the three variables in commonSettings.cmake and leave the rest as it is (maybe removing boost/test/test_tools.hpp as this caused a hidden compiler error)

What about the minimum boost version, should this be 1_72 or 1_69 (below I couldn't build ORE at all)?

**Comment by rkapl123** [comment]
OK, I've decided to do the minimal possible change in #204, because adding `BOOST_USE_WINAPI_VERSION=0x0A00` on a ORE global level made a mess with building Quantlib/QuantExt. Thanks for everyone helping on this!
